x-pack/plugins/ingest_manager/server/services/epm/archive/index.ts

```diff
@@ -0,0 +1,320 @@
+    );
+  }
+
+  const { name, version, description, type, categories, format_version: formatVersion } = manifest;
+  // check for mandatory fields
+  if (!(name && version && description && type && categories && formatVersion)) {
+    throw new PackageInvalidArchiveError(
+      'Invalid top-level package manifest: one or more fields missing of name, version, description, type, categories, format_version'
+    );
+  }
+
+  const datasets = parseAndVerifyDatasets(paths, name, version);
+  const configTemplates = parseAndVerifyConfigTemplates(manifest);
+
+  return {
+    name,
+    version,
+    description,
+    type,
+    categories,
+    format_version: formatVersion,
+    datasets,
+    config_templates: configTemplates,
+  };
+}
+
+function parseAndVerifyDatasets(paths: string[], pkgName: string, pkgVersion: string): Dataset[] {
+  // A dataset is made up of a subdirectory of name-version/dataset/, containing a manifest.yml
+  let datasetPaths: string[] = [];
+  const datasets: Dataset[] = [];
+
+  // pick all paths matching name-version/dataset/DATASET_PATH/...
+  // from those, pick all unique dataset paths
+  paths
+    .filter((path) => path.startsWith(`${pkgName}-${pkgVersion}/dataset/`))
+    .forEach((path) => {
+      const parts = path.split('/');
+      if (parts.length > 2 && parts[2]) datasetPaths.push(parts[2]);
+    });
+
+  datasetPaths = uniq(datasetPaths);
+
+  datasetPaths.forEach((datasetPath) => {
+    const manifestFile = `${pkgName}-${pkgVersion}/dataset/${datasetPath}/manifest.yml`;
+    const manifestBuffer = cacheGet(manifestFile);
+    if (!paths.includes(manifestFile) || !manifestBuffer) {
+      throw new PackageInvalidArchiveError(
+        `No manifest.yml file found for dataset '${datasetPath}'`
+      );
+    }
+
+    let manifest;
+    try {
+      manifest = yaml.load(manifestBuffer.toString());
+    } catch (error) {
+      throw new PackageInvalidArchiveError(
+        `Could not parse package manifest for dataset '${datasetPath}': ${error}.`
+      );
+    }
+
+    const { title: datasetTitle, release, ingest_pipeline: ingestPipeline, type } = manifest;
+    if (!(datasetTitle && release && type)) {
+      throw new PackageInvalidArchiveError(
+        `Invalid manifest for dataset '${datasetPath}': one or more fields missing of 'title', 'release', 'type'`
+      );
+    }
+    const streams = parseAndVerifyStreams(manifest, datasetPath);
+
+    // default ingest pipeline name see https://github.com/elastic/package-registry/blob/master/util/dataset.go#L26
+    return datasets.push({
+      name: `${pkgName}.${datasetPath}`,
+      title: datasetTitle,
+      release,
+      package: pkgName,
+      ingest_pipeline: ingestPipeline || 'default',
+      path: datasetPath,
+      type,
+      streams,
+    });
+  });
+
+  return datasets;
+}
+
+function parseAndVerifyStreams(manifest: any, datasetPath: string): RegistryStream[] {
```

Comment:
Would it simplify things if we had the type for manifest defined?

Reply:
At this point, `manifest` is an object that comes from the parsed YAML. I could type it, but remember that TypeScript does not perform run-time verification, so we still need code to verify that it contains all mandatory fields at run time.

The `parseAndVerify` methods all return correctly typed objects, and I could use the existing types that we already use for registry packages. (They still need to be refactored out of the `Registry*` namespace as they apply to all `InstallablePackage`s now.)

Also, This code will (probably? hopefully?) be replaced by "proper" verification based on https://github.com/elastic/package-spec/ .

Reply:
> Also, this code will (probably? hopefully?) be replaced by "proper" verification based on https://github.com/elastic/package-spec/ .

I believe so. The spec is written in JSON Schema, so we can use [AJV](https://github.com/ajv-validator/ajv) to validator functions, like we do for some of the agent routes https://github.com/elastic/kibana/blob/master/x-pack/plugins/ingest_manager/server/routes/agent/index.ts#L67

Reply:
Also, there's also a Go validator implementation for the package spec here: https://github.com/elastic/package-spec/tree/master/code/go

Maybe we can add one in TypeScript and put it into a npm module, that would put the spec and the validator a bit closer together.

@ycombinator do you have plans for a TypeScript implementation of the package-spec validator?

Reply:
Yes, that was exactly the intent - to have validators in multiple languages, depending on where they were needed. It’s why we have the folder structure we do in the package-spec repo. 

Unfortunately I’m not well-versed in Typescript so it would be great if someone fluent in it could send PRs to the package-spec repo.

Reply:
@ycombinator I'll look into it.

Reply:
@skh I'm happy to help as well. LMK if you need anything. I think AJV will do the work for us. They have a beta release which supports TS

Reply:
@jfsiii I have pinged you on https://github.com/elastic/kibana/issues/79432 , happy to collaborate on this!